CORE-10051: rename HSMServiceImpl and rename to TenantInfoService, without OSGi; eliminate HSMStore and CryptoServiceFactory

dickon opened the pull request:
CORE-10051: remove HSMServiceImpl OSGi, and rename to TenantInfoService

Remove HSMStore; use HSMRepository directly instead.

Remove hsmId as far as  possible, since it only ever has one value ("SOFT").

move CryptoService to from libs/crypto/cipher-suite to  libs/crypto/crypto-core so that it can use declarations in crypto-core

dickon committed:
CORE-10051: remove HSMServiceImpl OSGi, and rename to TenantInfoService

Remove HSMStore; use HSMRepository directly instead.

Remove hsmId wherever possible, since it only ever has one value ("SOFT")

remove import

chrisr3 commented:
This is inevitably going to impact the "crypto" code in the driver. Can we merge #4251 and #4254 before this one please? Then I can merge `release/os/5.1` into the `feature/flow-testing` branch and defer updating the driver's crypto integration until you have completed your changes.

dickon committed:
rename private method to startTenantInfoService

dickon committed:
fix: remove old ref

simon-johnson-r3 commented on components/crypto/crypto-service-impl/src/main/kotlin/net/corda/crypto/service/TenantInfoService.kt:
```diff
@@ -0,0 +1,25 @@
+    /**
+     * Populate settings for a specific category on a specific tenant  
+     * 
+     * Result type is HSMAssociationInfo, which is a type defined in the public API and therefore
```
This comment wont make sense out of the context of this PR, e.g. 6 months from now. I think it would be better to explain it in terms of the standard
`@return Information pertinent to the tenant registration`

And then maybe change the description of `HSMAssociationInfo` over in that class file to suggest some of the information might not be useful at this point in time.

dickon replied:
`HSMAssociationInfo` is an autogenerated Avro stub, so we cannot include our own javadoc. It's also public API so I'm not looking to change at this stage

dickon replied:
Done. Comments rewritten to kdoc conventions and be more self-contained.

That is good pracitce and Corda standard, even though we violate that in places.

simon-johnson-r3 commented on components/crypto/crypto-service-impl/src/main/kotlin/net/corda/crypto/service/TenantInfoService.kt:
```diff
@@ -0,0 +1,25 @@
+    fun populate(tenantId: String, category: String): HSMAssociationInfo
+
+    /**
+     * Lookup settings assigned HSM by tenant and category.
```
I think there's words missing here, or am I reading it incorrectly?

dickon replied:
fixed

simon-johnson-r3 commented on components/crypto/crypto-service-impl/src/main/kotlin/net/corda/crypto/service/impl/SigningServiceImpl.kt:
```diff
@@ -320,23 +293,31 @@ class SigningServiceImpl(
 
-            val generatedKey = ref.instance.generateKeyPair(
-                KeyGenerationSpec(scheme, alias, ref.masterKeyAlias),
+            // TODO always return GeneratedWrappedKey here
```
Can we remove this TODO, perhaps by creating a new ticket? It's not obvious when it will be done. If it's already in a later PR, perhaps we can just not mention it?

dickon replied:
Done. It turns out I'd already made the change so the TODO was an error.

simon-johnson-r3 commented on components/crypto/crypto-service-impl/src/test/kotlin/net/corda/crypto/service/impl/SigningServiceGeneralTests.kt:
```diff
@@ -55,6 +57,7 @@ class SigningServiceGeneralTests {
     companion object {
         private lateinit var schemeMetadata: CipherSchemeMetadata
 
+        private val wrappingKeyAlias = "Enoch Root"
```
Not sure what the policy on stuff like this is. Given this is an open source project I suggest we try to avoid it and stick to something formal like "alias1"?

dickon replied:
done

simon-johnson-r3 commented on components/crypto/crypto-service-impl/src/test/kotlin/net/corda/crypto/service/impl/SigningServiceImplTest.kt:
```diff
@@ -4,24 +4,36 @@ import net.corda.crypto.cipher.suite.CipherSchemeMetadata
-    private val cryptoServiceFactory = mock<CryptoServiceFactory>()
     private val schemeMetadata = mock<CipherSchemeMetadata>()
+
+    private val masterKeyAlias = "IAMGROOT"
```
same here

dickon replied:
done

simon-johnson-r3 commented on components/crypto/crypto-service-impl/src/test/kotlin/net/corda/crypto/service/impl/infra/TestServicesFactory.kt:
```diff
@@ -200,27 +168,27 @@ class TestServicesFactory {
-            assertEquals(LifecycleStatus.UP, it.lifecycleCoordinator.status)
-        }
+    val signingConfig = cryptoConfig.signingService()
+    val signingService: SigningService by lazy {
```
is there a reason this is lazy?

dickon replied:
To speed up test run times. Not all test classes that use this suite of test fixtures need all of them. We see this in other parts of Corda e.g. P2P.

The fact that not all of them are lazy is a bit weird, but for the most part the more expensive constructors are handled via `lazy`.

simon-johnson-r3 commented on components/crypto/crypto-softhsm-impl/src/main/kotlin/net/corda/crypto/softhsm/impl/SoftCryptoService.kt:
```diff
@@ -45,6 +45,8 @@ const val PRIVATE_KEY_ENCODING_VERSION: Int = 1
  *
  * @param wrappingRepositoryFactory which provides a factory for [WrappingRepository], which provides save and
  *        find operations for wrapping keys on a specific tenant.
+ * @param signingRepositoryFactory which provides a factory for [SigningRepository], which provides save and
```
There's no `signingRepositoryFactory` parameter here.

dickon replied:
fixed

simon-johnson-r3 commented on components/crypto/crypto-softhsm-impl/src/test/kotlin/net/corda/crypto/softhsm/impl/SoftCryptoServiceOperationsTests.kt:
```diff
@@ -86,6 +88,7 @@ class SoftCryptoServiceOperationsTests {
             )
         )
         private val wrappingKeyCache = makeWrappingKeyCache()
+        private val signingRepository: SigningRepository = mock()
```
is this used anywhere?

dickon replied:
deleted

simon-johnson-r3 commented on components/crypto/crypto-softhsm-impl/src/test/kotlin/net/corda/crypto/softhsm/impl/infra/MakeCryptoService.kt:
```diff
@@ -22,7 +22,7 @@ fun makeSoftCryptoService(
         WrappingKeyImpl.generateWrappingKey(it)
     },
-    wrappingRepository: WrappingRepository = TestWrappingRepository(),
+    wrappingRepository: WrappingRepository = TestWrappingRepository()
```
Just a note that this trailing comma is actually more aligned to the Kotlin coding standards than not having it. I don't suggest you add it back in now, but I don't recommend removing them in the future either (our code base is split between the two at the moment).

dickon replied:
We need a consensus on this between you and @chrisr3 who has pushed the opposite way.

However, it wasn't intentional so I reverted it.

simon-johnson-r3 commented on libs/crypto/crypto-config-impl/src/main/kotlin/net/corda/crypto/config/impl/CryptoConfigUtils.kt:
```diff
@@ -161,6 +171,7 @@ fun SmartConfig.toConfigurationSecrets(): ConfigurationSecrets = ConfigurationSe
 fun SmartConfig.signingService(): CryptoSigningServiceConfig =
     CryptoSigningServiceConfig(this)
 
+// TODO remove me, dead?
```
I would either remove the method or the comment. It will build or it wont after all!

dickon replied:
TODO removed. It is still used.

simon-johnson-r3 commented on libs/crypto/crypto-config-impl/src/main/kotlin/net/corda/crypto/config/impl/CryptoConfigUtils.kt:
```diff
@@ -142,14 +142,24 @@ import net.corda.schema.configuration.ConfigKeys.CRYPTO_CONFIG
 }
  */
 
-private const val HSM_ID = "hsmId"
```
if these were private before, are they used anywhere?

dickon replied:
deleted it instead of making it public

simon-johnson-r3 commented on libs/crypto/crypto-config-impl/src/main/kotlin/net/corda/crypto/config/impl/CryptoHSMConfig.kt:
```diff
@@ -3,6 +3,7 @@ package net.corda.crypto.config.impl
 import com.typesafe.config.ConfigList
 import net.corda.libs.configuration.SmartConfig
 
+// TODO remove me, dead?
```
same, I would either remove the comment or the class

dickon replied:
TODO deleted. At the time I was trying to absolutely minimise PR size, so deleting more at once than necessary goes against that desire to mimise changes.

simon-johnson-r3 commented on components/crypto/crypto-service-impl/src/main/kotlin/net/corda/crypto/service/impl/SigningServiceImpl.kt:
```diff
@@ -1,28 +1,28 @@
 
 import com.github.benmanes.caffeine.cache.Cache
-import com.github.benmanes.caffeine.cache.Caffeine
-import net.corda.cache.caffeine.CacheFactoryImpl
```
might be used elsewhere in this module, but if not, can Caffeine be removed from the build.gradle? (triggered by seeing it included in another build.gradle below)

dickon replied:
No, we still have a reference to `com.github.benmanes.caffeine.cache.Cache` here, and we construct caffeine in test cases. However, api+testImplementaiton should work so I'll try that.

simon-johnson-r3 commented on processors/crypto-processor/src/main/kotlin/net/corda/processors/crypto/internal/CryptoProcessorImpl.kt:
```diff
@@ -172,66 +183,111 @@ class CryptoProcessorImpl @Activate constructor(
+    private fun startCryptoService(config: SmartConfig) {
+        logger.info("Creating instance of the {}", SoftCryptoService::class.java.name)
+        val cachingConfig = config.getConfig(CACHING)
+        val expireAfterAccessMins = cachingConfig.getConfig(EXPIRE_AFTER_ACCESS_MINS).getLong(DEFAULT)
```
I know it comes from elsewhere, but I find `DEFAULT` a bit of a vague constant name...

dickon replied:
Maybe so, but I don't think we change or should change this now; this is code which has been moved rather than written for this purpose, and the names are in the config which is locked down public API .

simon-johnson-r3 commented on processors/crypto-processor/src/main/kotlin/net/corda/processors/crypto/internal/CryptoProcessorImpl.kt:
```diff
@@ -172,66 +183,111 @@ class CryptoProcessorImpl @Activate constructor(
+        require(unmanagedWrappingKeys.containsKey(defaultUnmanagedWrappingKeyName)) {
+            "default key $defaultUnmanagedWrappingKeyName must be in $HSM.$WRAPPING_KEYS"
+        }
+        // TODO drop this compatibility code that supports name between KEY_MAP_TRANSIENT_NAME and if so disables caching
```
does this need a ticket rather than a comment?

dickon replied:
This comment already exists in the code in its old location. However the TODO is old and has effectively been done so I'll remove  it.

Sakpal commented on components/crypto/crypto-softhsm-impl/src/main/kotlin/net/corda/crypto/softhsm/impl/HSMRepositoryImpl.kt:
```diff
@@ -74,13 +74,12 @@ class HSMRepositoryImpl(
-                findHSMAssociationEntity(em, tenantId, hsmId)
-                    ?: createAndPersistAssociation(em, tenantId, hsmId, masterKeyPolicy)
+                findHSMAssociationEntity(em, tenantId)
+                    ?: createAndPersistAssociation(em, tenantId, "SOFT", masterKeyPolicy)
```
Minor: Can we make a variable out of this string `"SOFT"` as I can see it has been used at multiple places?

dickon replied:
Done.

There already was a `val` in CryptConsts, I now use that everywhere.

dickon committed:
Use API and testing reference to caffeine

Improve kdocs on new TenantInfoService interface.

Drop some old TODOs that were implemented.

Remove parameter from kdoc which is no longer in the constructor

Undo trailing comma removal

Optimise imports.
Remove unused constant.

dickon reviewed (commented):
Thanks Simon for your detailed attention, I've addressed all that feedback.

dickon commented:
> This is inevitably going to impact the "crypto" code in the driver. Can we merge #4251 and #4254 before this one please? Then I can merge `release/os/5.1` into the `feature/flow-testing` branch and defer updating the driver's crypto integration until you have completed your changes.

These are both merged now.

seanbrereton reviewed (approved):
LGTM (blt)

yift-r3 reviewed (approved):
Network changes LGTM.

kyriathar commented on components/crypto/crypto-service-impl/src/main/kotlin/net/corda/crypto/service/HSMService.kt:
```diff
@@ -1,19 +0,0 @@
-/**
- * Internal HSM service to support all HSM configuration and assignment operations.
- */
-interface HSMService : Lifecycle {
```
FWIW I believe this change here makes this service not aligned anymore with the the rest of the SOFT hsm concept within the code base/ in the DB/ in rest APIs.

kyriathar commented on components/crypto/crypto-service-impl/src/main/kotlin/net/corda/crypto/service/TenantInfoService.kt:
```diff
@@ -0,0 +1,37 @@
+ * each tenant and category combination. Potentially one day we could store information about external key storage here.
+ */
+
+interface TenantInfoService {
+    /**
+     * Populate settings for a specific category on a specific tenant, and return those settings.
+     *
+     * @param tenantId the tenant (virtual node or special purpose cluster level
+     * @param category indicates what the keys will be used for, and should be one of the constants
+     *                 defined in CryptoConsts.Categories
+     *
+     * @return Information about crypto processing for a specific category of keys in a specific tenant.
+     *         Includes `masterKeyAlias` which identifies the per-tenant master key used to wrap keys.
+     */
+    fun populate(tenantId: String, category: String): HSMAssociationInfo
+
+    /**
+     * Lookup settings for a specific category on a specific tenant.
+     *
+     * @param tenantId the tenant (virtual node or special purpose cluster level
+     * @param category indicates what the keys will be used for, and should be one of the constants
+     *                 defined in CryptoConsts.Categories
+     *
+     * @return Information about crypto processing for a specific category of keys in a specific tenant.
+     *         Includes `masterKeyAlias` which identifies the per-tenant master key used to wrap keys.
+     */
+
+    fun lookup(tenantId: String, category: String): HSMAssociationInfo?
```
FWIW I believe these APIs here are now less meaningful of what their actually doing, compared to what they used to be before.
